File: frontend/src/api/types/requirement.ts
export type RequirementType = 'core' | 'non_core'

export type RequirementStatus = 'draft' | 'active' | 'done' | 'cancelled'

export type RequirementPriority = 'low' | 'medium' | 'high' | 'urgent'

export interface RequirementOwner {
  id: number
  username: string
  display_name: string
}

export interface Requirement {
  id: number
  project_id: number
  parent_id: number | null
  title: string
  description: string | null
  type: RequirementType
  priority: RequirementPriority | null
  status: RequirementStatus
  owner_id: number | null
  owner?: RequirementOwner | null
  cancelled_at: string | null
  cancel_reason: string | null
  created_at: string
  updated_at: string
}

export interface RequirementTreeNode extends Requirement {
  children?: RequirementTreeNode[]
}

export interface RequirementListParams {
  tree?: boolean
  type?: RequirementType
  skip?: number
  limit?: number
}

export interface RequirementListResponse {
  items: Requirement[]
  total: number
  skip: number
  limit: number | null
}

export interface RequirementTreeListResponse {
  items: RequirementTreeNode[]
  total: number
  skip: number
  limit: number | null
}

export interface RequirementCreatePayload {
  title: string
  description?: string | null
  type: RequirementType
  priority?: RequirementPriority | null
  parent_id?: number | null
  owner_id?: number | null
  status?: RequirementStatus
}

export interface RequirementUpdatePayload {
  title?: string
  description?: string | null
  type?: RequirementType
  priority?: RequirementPriority | null
  parent_id?: number | null
  owner_id?: number | null
  status?: RequirementStatus
}

export interface RequirementCancelPayload {
  cancel_reason: string
}

export const REQUIREMENT_TYPE_LABELS: Record<RequirementType, string> = {
  core: '核心业务',
  non_core: '非核心业务',
}

export const REQUIREMENT_STATUS_LABELS: Record<RequirementStatus, string> = {
  draft: '草稿',
  active: '进行中',
  done: '已完成',
  cancelled: '已取消',
}

export const REQUIREMENT_PRIORITY_LABELS: Record<RequirementPriority, string> = {
  low: '低',
  medium: '中',
  high: '高',
  urgent: '紧急',
}

/** 末节点：不存在以本需求为 parent 的子需求 */
export function isRequirementLeaf(
  requirementId: number,
  requirements: Requirement[],
): boolean {
  return !requirements.some((item) => item.parent_id === requirementId)
}

File: backend/app/schemas/requirement.py
from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.requirement import (
    RequirementPriority,
    RequirementStatus,
    RequirementType,
)


class RequirementCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: Optional[str] = None
    type: RequirementType
    priority: Optional[RequirementPriority] = None
    parent_id: Optional[int] = None
    owner_id: Optional[int] = None
    status: RequirementStatus = RequirementStatus.DRAFT


class RequirementUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = None
    type: Optional[RequirementType] = None
    priority: Optional[RequirementPriority] = None
    parent_id: Optional[int] = None
    owner_id: Optional[int] = None
    status: Optional[RequirementStatus] = None


class RequirementCancel(BaseModel):
    cancel_reason: str = Field(min_length=1)


class RequirementOwnerBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    username: str
    display_name: str


class RequirementResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    parent_id: Optional[int]
    title: str
    description: Optional[str]
    type: RequirementType
    priority: Optional[RequirementPriority]
    status: RequirementStatus
    owner_id: Optional[int]
    owner: Optional[RequirementOwnerBrief] = None
    cancelled_at: Optional[datetime]
    cancel_reason: Optional[str]
    created_at: datetime
    updated_at: datetime


class RequirementTreeNode(RequirementResponse):
    children: List[RequirementTreeNode] = []


class RequirementListResponse(BaseModel):
    items: List[RequirementResponse]
    total: int
    skip: int
    limit: Optional[int]


class RequirementTreeListResponse(BaseModel):
    items: List[RequirementTreeNode]
    total: int
    skip: int
    limit: Optional[int]


RequirementTreeNode.model_rebuild()
